express-notes-api/server.ts

```diff
@@ -0,0 +1,154 @@
+import express from 'express';
+import { readFile, writeFile } from 'node:fs/promises';
+
+type Note = {
+  id: number;
+  content: string;
+};
+
+type Data = {
+  nextId: number;
+  notes: Record<number, Note>;
+};
+let data: Data;
+
+type Err = {
+  error: string;
+};
+
+const error: Err = {
+  error: 'An error occured.',
+};
+
+const app = express();
+app.use(express.json());
+
+// Returns list of notes in JSON
+app.get('/api/notes', async (req, res) => {
+  try {
+    await readJSON();
+    const notesList: Note[] = [];
+    for (const key in data.notes) {
+      notesList.push(data.notes[key]);
+    }
+    res.status(200).json(notesList);
+  } catch (err) {
+    console.error(err);
+  }
+});
+
+// Returns individual note based on id in JSON
+app.get('/api/notes/:id', async (req, res) => {
+  try {
+    await readJSON();
+    const id: number = Number(req.params.id);
+    if (isNaN(id) || id < 1 || !Number.isInteger(id)) {
+      error.error = `${req.params.id} is not a valid id number.`;
+      res.status(400).json(error);
+      throw new Error(`${req.params.id} is not a valid id number.`);
```

Comment:
Oh, I see why your catch block doesn't send a response -- it's so you can throw here and have it logged in the catch block.

Unfortunately, this doesn't quite work. The reason is because if `readJSON` or anything else in here throws an exception, it will go straight to catch and the client won't receive a good response.

We'll learn a better way to do error handling in the next exercise, but I suggest you simply `return` here and add a 500 response in the catch clause.

Note also that if the client makes a bad request (so you send a 400 response), you don't want to log it. You only want to log things that will help you in debugging your server code. 400 errors, though, are client problems, so you don't usually log those. (Otherwise a malicious client could send you thousands of bad requests and fill up your log file, potentially causing the server to crash.)